### PR TITLE
fix!: Toby/perf

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/core/model/definition.py [hunk 1]
```diff
@@ -1202,6 +1236,17 @@ class SeedModel(_SqlBasedModel):
...
 
+    def __getstate__(self) -> t.Dict[t.Any, t.Any]:
+        state = super().__getstate__()
+        state["__dict__"] = state["__dict__"].copy()
+        state["__dict__"].pop("_reader", None)
+        return state
+
+    def copy(self, **kwargs: t.Any) -> Self:
+        model = super().copy(**kwargs)
+        model.__dict__.pop("_reader", None)
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1247]: What's `_reader`?

File: sqlmesh/core/context.py [hunk 2]
```diff
@@ -427,12 +426,20 @@ def upsert_model(self, model: t.Union[str, Model], **kwargs: t.Any) -> Model:
             raise SQLMeshError(f"The disabled model '{model.name}' cannot be upserted")
         path = model._path
 
-        # model.copy() can't be used here due to a cached state that can be a part of a model instance.
-        model = t.cast(Model, type(model)(**{**t.cast(Model, model).dict(), **kwargs}))
+        model = model.copy(update=kwargs)
+        model._full_depends_on = None  # the query can be updated to change deps
```
Comment (izeigerman) [inline sqlmesh/core/context.py:430]: Why do we not do this in the model's `copy()`?

File: sqlmesh/core/model/definition.py (hunk 1 above)
Comment (tobymao) [reply sqlmesh/core/model/definition.py:1247]: seed reader / csv reader

File: sqlmesh/core/context.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/core/context.py:430]: because most of the times you want to cache _full_depends_on, it's just in context.upsert (like in tests), we mutate the query which can change depends_on

Comment (izeigerman) [reply sqlmesh/core/context.py:430]: I think this may lead to too many gotchas. I'd check in `copy` whether `update` include `depends_on` or `query`. And then invalidate based on that. Otherwise this is too error prone.

Comment (izeigerman) [reply sqlmesh/core/context.py:430]: Also I don't see us resetting renderers on copy

File: sqlmesh/core/model/cache.py [hunk 3]
```diff
@@ -44,22 +42,20 @@ def get_or_load(self, name: str, entry_id: str = "", *, loader: t.Callable[[], M
...
-        loaded_model = loader()
-        if isinstance(loaded_model, SqlModel):
-            new_entry = SqlModelCacheEntry(
-                model=loaded_model, full_depends_on=loaded_model.full_depends_on
-            )
-            self._file_cache.put(name, entry_id, value=new_entry)
+        model = loader()
+        if isinstance(model, SqlModel):
+            # make sure we preload full_depends_on
+            model.full_depends_on
```
Comment (izeigerman) [inline sqlmesh/core/model/cache.py:51]: Such a hack :) This is why `full_depends_on` was a separate attribute - to explicitly store it and not rely on internal mechanics of materializing the cache.

File: sqlmesh/core/model/definition.py [hunk 4]
```diff
@@ -698,7 +711,7 @@ def is_python(self) -> bool:
     def is_seed(self) -> bool:
         return False
 
-    @cached_property
+    @property
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:714]: Why this changed?

File: sqlmesh/core/model/definition.py [hunk 5]
```diff
@@ -995,6 +1012,23 @@ class SqlModel(_SqlBasedModel):
...
+        # query renderer is very expensive to serialize
+        state["__dict__"].pop("_query_renderer", None)
+        state["__dict__"].pop("column_descriptions", None)
+        private = state[PRIVATE_FIELDS]
+        private["_columns_to_types"] = None
+        return state
+
+    def copy(self, **kwargs: t.Any) -> Self:
+        model = super().copy(**kwargs)
+        model.__dict__.pop("_query_renderer", None)
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1027]: Ah, found it. Why not reset it using the attribute directly? Why pop it?

File: sqlmesh/core/node.py [hunk 6]
```diff
@@ -193,6 +200,8 @@ class _Node(PydanticModel):
     tags: t.List[str] = []
     stamp: t.Optional[str] = None
     _path: Path = Path()
+    _data_hash: t.Optional[str] = None
```
Comment (izeigerman) [inline sqlmesh/core/node.py:203]: Why is this here and not in the `_Model`? I think it only applies there.

File: sqlmesh/core/state_sync/engine_adapter.py [hunk 7]
```diff
@@ -1425,6 +1427,7 @@ def _snapshot_to_json(snapshot: Snapshot) -> str:
 
 def parse_snapshot(
     model_cache: ModelCache,
+    optimized_query_cache: OptimizedQueryCache,
```
Comment (izeigerman) [inline sqlmesh/core/state_sync/engine_adapter.py:1430]: Let's not do this here. I'm going to work on a more comprehensive snapshot caching next.

File: sqlmesh/core/model/definition.py (hunk 5 above)
Comment (tobymao) [reply sqlmesh/core/model/definition.py:1027]: because it's a cached_property and it's behavior is special

File: sqlmesh/core/node.py (hunk 6 above)
Comment (tobymao) [reply sqlmesh/core/node.py:203]: no, audits also have these properties

File: sqlmesh/core/model/definition.py (hunk 4 above)
Comment (tobymao) [reply sqlmesh/core/model/definition.py:714]: because it's annoying to deal with cached_properties in cleanup

File: sqlmesh/core/context.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/core/context.py:430]: happens when you to context.upsert_model, used in tests, and notebooks

File: sqlmesh/core/model/definition.py [hunk 8]
```diff
@@ -995,6 +1012,25 @@ class SqlModel(_SqlBasedModel):
...
+        model.__dict__.pop("column_descriptions", None)
+        model._columns_to_types = None
+        if "query" in kwargs.get("update", {}):
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1030]: We should check for `depends_on` here too.

**Review by izeigerman — APPROVED** [review]
